python/cudf/cudf/dataframe/multiindex.py

```diff
@@ -204,31 +204,42 @@ def _compute_levels_and_codes(self):
         self._levels = levels
         self._codes = codes
 
-    def _compute_validity_mask(self, df, row_tuple):
+    def _compute_validity_mask(self, index, row_tuple):
         """ Computes the valid set of indices of values in the lookup
         """
-        validity_mask = []
-        for i, element in enumerate(row_tuple):
-            index_of_code_at_level = None
-            for level_index in range(len(self.levels[i])):
-                if self.levels[i][level_index] == element:
-                    index_of_code_at_level = level_index
-                    break
-            if index_of_code_at_level is None:
-                raise KeyError(element)
-            matches = []
-            for k, code in enumerate(self.codes[self.codes.columns[i]]):
-                if k in validity_mask or len(validity_mask) == 0:
-                    if code == index_of_code_at_level:
-                        matches.append(k)
-            if len(matches) != 0:
-                validity_mask = matches
-        return validity_mask
+        # Instructions for Slicing
+        # if tuple, get first and last elements of tuple
+        # if open beginning tuple, get 0 to highest valid_index
+        # if open ending tuple, get highest valid_index to len()
+        # if not open end or beginning, get range lowest beginning index
+        # to highest ending index
+        from cudf import DataFrame
+        from cudf import Series
+        from cudf import concat
+        from cudf.utils.cudautils import arange
+        lookup = DataFrame()
+        if isinstance(row_tuple, slice):
+            start, stop, step, sln = utils.standard_python_slice(
+                len(index.codes),
+                row_tuple
+            )
+            return Series(arange(start, stop, step))
+        for idx, row in enumerate(row_tuple):
+            if row == slice(None):
+                continue
+            try:
+                lookup[index.codes.columns[idx]] = Series(
```

Comment:
Should be able to use `_source_data` here instead of codes to prevent materialization.

Reply:
I don't think that using codes can be completely avoided in this function - the goal of `_compute_validity_mask` is to find all of the rows (or columns) that match a set of lookup keys that map into levels, for their index value, and codes, for the row (or column) that matches that particular index level. Is there a way to find out which rows match a particular index tuple without knowing the codes of that row, which depend on the levels? I don't think so.

Reply:
This is an example of why I don't think `codes` can be avoided:

https://www.dropbox.com/s/foq2qyss8d56y05/Screenshot%202019-07-11%2011.05.01.png?raw=1